File: front_end/lib/service/gerenciar_lista/item_service.dart
import 'package:crud_flutter/core/api/api_client.dart';
import 'package:crud_flutter/dto/request/gerenciar_lista/adicionar_produto_lista_request_dto.dart';
import 'package:crud_flutter/dto/request/gerenciar_lista/item_request_create_dto.dart';
import 'package:crud_flutter/dto/request/gerenciar_lista/item_request_update_dto.dart';
import 'package:crud_flutter/model/gerenciar_lista/item.dart';

class ItemService {
  final ApiClient _client;

  ItemService(this._client);

  // =====================================================
  // 📥 LISTAR ITENS
  // =====================================================
  Future<List<Item>> listar(int listaId) async {
    final url = '/listas/$listaId/itens';

    final result = await _client.get<List<Item>>(
      url,
      (data) => (data as List).map((e) => Item.fromJson(e)).toList(),
    );

    if (!result.success) {
      throw Exception(result.message);
    }

    return result.data ?? [];
  }

  // =====================================================
  // ➕ CRIAR ITEM
  // =====================================================
  Future<Item> criar(int listaId, ItemCreateDTO dto) async {
    final url = '/listas/$listaId/itens';

    final result = await _client.post<Item>(
      url,
      dto.toJson(),
      (data) => Item.fromJson(data),
    );

    if (!result.success || result.data == null) {
      throw Exception(result.message);
    }

    return result.data!;
  }

  // =====================================================
  // ✏️ ATUALIZAR ITEM
  // =====================================================
  Future<Item> atualizar(
    int listaId,
    int idItem,
    ItemUpdateDTO dto,
  ) async {
    final url = '/listas/$listaId/itens/$idItem';

    final result = await _client.put<Item>(
      url,
      dto.toJson(),
      (data) => Item.fromJson(data),
    );

    if (!result.success || result.data == null) {
      throw Exception(result.message);
    }

    return result.data!;
  }

  // =====================================================
  // ➕ ADICIONAR PRODUTO NA LISTA
  // =====================================================
  Future<Item> adicionarProdutoNaLista(
    int listaId,
    AdicionarProdutoListaRequestDTO dto,
  ) async {
    final url = '/listas/$listaId/itens';

    final result = await _client.post<Item>(
      url,
      dto.toJson(),
      (data) => Item.fromJson(data),
    );

    if (!result.success || result.data == null) {
      throw Exception(result.message);
    }

    return result.data!;
  }

  // =====================================================
  // ✅ MARCAR COMO COMPRADO
  // =====================================================
  Future<void> marcarComprado(int listaId, int idItem) async {
    final url = '/listas/$listaId/itens/$idItem/comprado';

    print("📡 [ITEM SERVICE] marcarComprado INICIO");
    print("📡 [ITEM SERVICE] listaId=$listaId itemId=$idItem");
    print("🌐 [ITEM SERVICE] URL=$url");

    try {
      final result = await _client.patch<void>(
        url,
        {},
        null,
      );

      print("📨 [ITEM SERVICE] resposta recebida");
      print("📨 [ITEM SERVICE] success=${result.success}");
      print("📨 [ITEM SERVICE] message=${result.message}");

      if (!result.success) {
        print("❌ [ITEM SERVICE] ERRO ao marcar como comprado");
        throw Exception(result.message);
      }

      print("✅ [ITEM SERVICE] item marcado como comprado com sucesso");
    } catch (e) {
      print("💥 [ITEM SERVICE] EXCEPTION: $e");
      rethrow;
    }
  }

  // =====================================================
  // ❌ DESMARCAR COMO COMPRADO
  // =====================================================
  Future<void> desmarcarComprado(int listaId, int idItem) async {
    final url = '/listas/$listaId/itens/$idItem/desmarcar';

    final result = await _client.patch<void>(
      url,
      {},
      null,
    );

    if (!result.success) {
      throw Exception(result.message);
    }
  }

  // =====================================================
  // 🗑 DELETAR ITEM
  // =====================================================
  Future<void> deletar(int listaId, int idItem) async {
    final url = '/listas/$listaId/itens/$idItem';

    final result = await _client.delete<void>(url);

    if (!result.success) {
      throw Exception(result.message);
    }
  }
}

File: back-end/src/main/java/com/tnl/listacompras/service/gerenciar_lista/ItemService.java
package com.tnl.listacompras.service.gerenciar_lista;

import com.tnl.listacompras.dto.requestDTO.gerenciar_lista.ItemRequestDTO;
import com.tnl.listacompras.dto.requestDTO.gerenciar_lista.ItemUpdateDTO;
import com.tnl.listacompras.dto.responseDTO.gerenciar_lista.ItemResponseDTO;
import com.tnl.listacompras.model.cadastrar_produto.Produto;
import com.tnl.listacompras.model.gerenciar_lista.Item;
import com.tnl.listacompras.model.gerenciar_lista.Lista;
import com.tnl.listacompras.repository.cadastrar_produto.ProdutoRepository;
import com.tnl.listacompras.repository.gerenciar_lista.ItemRepository;

import exception.business.BusinessException;
import exception.business.NotFoundException;

import org.springframework.stereotype.Service;

import java.util.List;

@Service
public class ItemService {

    private final ItemRepository itemRepository;
    private final ProdutoRepository produtoRepository;

    public ItemService(
            ItemRepository itemRepository,
            ProdutoRepository produtoRepository
    ) {
        this.itemRepository = itemRepository;
        this.produtoRepository = produtoRepository;
    }

    // =========================
    // HELPER ITEM
    // =========================
    private Item validarItem(Long listaId, Long itemId) {

        return itemRepository
                .findByIdAndListaIdAndDeletadoFalse(itemId, listaId)
                .orElseThrow(() ->
                        new NotFoundException("Item não encontrado"));
    }

    // =========================
    // LISTAR
    // =========================
    public List<ItemResponseDTO> listarPorLista(Long listaId) {

        return itemRepository
                .findByListaIdAndDeletadoFalse(listaId)
                .stream()
                .map(ItemResponseDTO::new)
                .toList();
    }

    // =========================
    // CRIAR
    // =========================
    // TODO: Revisar criação de Lista aqui.
    // Atualmente estamos usando new Lista(listaId) apenas para referência.
    // Ideal seria garantir a entidade gerenciada (Lista real) vinda do ListaService
    // para evitar inconsistências no contexto do Hibernate e melhorar consistência do domínio.
    public ItemResponseDTO criar(Lista lista, ItemRequestDTO dto) {

        Produto produto = produtoRepository.findById(dto.getProdutoId())
                .orElseThrow(() -> new NotFoundException("Produto não encontrado"));

        if (lista.getConcluidoEm() != null) {
            throw new BusinessException("Lista já concluída");
        }

        Item item = itemRepository
                .findByListaIdAndProdutoIdAndDeletadoFalse(lista.getId(), produto.getId())
                .orElse(null);

        if (item != null) {
            item.setQuantidade(item.getQuantidade() + dto.getQuantidade());
            item.setPreco(dto.getPreco());
            return new ItemResponseDTO(itemRepository.save(item));
        }

        Item novo = new Item();
        novo.setLista(lista); // ✔ entidade REAL e gerenciada
        novo.setProduto(produto);
        novo.setQuantidade(dto.getQuantidade());
        novo.setPreco(dto.getPreco());

        return new ItemResponseDTO(itemRepository.save(novo));
    }

    // =========================
    // ATUALIZAR
    // =========================
    public ItemResponseDTO atualizar(
            Long listaId,
            Long itemId,
            ItemUpdateDTO dto
    ) {

        Item item = validarItem(listaId, itemId);

        if (dto.getQuantidade() == null || dto.getQuantidade() <= 0) {
            throw new BusinessException(
                    "Quantidade deve ser maior que zero"
            );
        }

        if (dto.getPreco() == null || dto.getPreco() < 0) {
            throw new BusinessException(
                    "Preço não pode ser negativo"
            );
        }

        item.setQuantidade(dto.getQuantidade());
        item.setPreco(dto.getPreco());

        return new ItemResponseDTO(itemRepository.save(item));
    }

    // =========================
    // MARCAR COMO COMPRADO
    // =========================
    public void marcarComprado(Long listaId, Long itemId) {

        Item item = validarItem(listaId, itemId);

        item.setComprado(true);

        itemRepository.save(item);
    }

    // =========================
    // DESMARCAR
    // =========================
    public void desmarcarComprado(Long listaId, Long itemId) {

        Item item = validarItem(listaId, itemId);

        item.setComprado(false);

        itemRepository.save(item);
    }

    // =========================
    // DELETE LÓGICO
    // =========================
    public void deletar(Long listaId, Long itemId) {

        Item item = validarItem(listaId, itemId);

        item.setDeletado(true);

        itemRepository.save(item);
    }

    // =========================
    // RESUMO
    // =========================
    public Long contarItensAtivos(Long listaId) {
        return itemRepository.contarItensAtivos(listaId);
    }

    public Long contarItensComprados(Long listaId) {
        return itemRepository.contarItensComprados(listaId);
    }
}